File: basic_dsa_three_hundred_forty_eight.java
/**
 * Basic DSA Problems - Design Tic-Tac-Toe (LeetCode 348)
 * Track row/col/diag sums for O(1) win detection on n x n board
 */
public class basic_dsa_three_hundred_forty_eight {

    // Problem: n x n board. move(row,col,player) returns player id if win, else 0.

    static class TicTacToe {
        int n;
        int[] rows, cols;
        int diag, antiDiag;

        TicTacToe(int n) {
            this.n = n;
            rows = new int[n];
            cols = new int[n];
        }

        public int move(int row, int col, int player) {
            int v = player == 1 ? 1 : -1;
            rows[row] += v;
            cols[col] += v;
            if (row == col) diag += v;
            if (row + col == n - 1) antiDiag += v;
            if (Math.abs(rows[row]) == n || Math.abs(cols[col]) == n
                    || Math.abs(diag) == n || Math.abs(antiDiag) == n)
                return player;
            return 0;
        }
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Design Tic-Tac-Toe");
        System.out.println("=".repeat(60));

        TicTacToe g = new TicTacToe(3);
        System.out.println("move(0,0,1)=" + g.move(0, 0, 1) + ", move(0,2,2)=" + g.move(0, 2, 2)
                + ", move(2,2,1)=" + g.move(2, 2, 1) + ", move(1,1,2)=" + g.move(1, 1, 2)
                + ", move(2,0,1)=" + g.move(2, 0, 1) + ", move(1,0,2)=" + g.move(1, 0, 2)
                + ", move(2,1,1)=" + g.move(2, 1, 1));
    }
}

File: basic_dsa_three_hundred_forty_eight.cpp
/**
 * Basic DSA - Fibonacci Number (LeetCode 509)
 * F(0)=0, F(1)=1, F(n)=F(n-1)+F(n-2).
 */
#include <iostream>
#include <string>

int fib(int n) {
    if (n <= 1) return n;
    int a = 0, b = 1;
    for (int i = 2; i <= n; i++) {
        int c = a + b;
        a = b;
        b = c;
    }
    return b;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Fibonacci Number\n";
    std::cout << std::string(60, '=') << "\n";

    std::cout << "2 -> " << fib(2) << "\n";
    std::cout << "3 -> " << fib(3) << "\n";
    std::cout << "4 -> " << fib(4) << "\n";
}
